solvebio/query.py

```diff
@@ -851,27 +861,41 @@ def join(self, query_b, key, key_b=None, prefix="b_", always_prefix=True):
         # Try to use a unique field for the sub-query data
         query_b_fields = query_b.fields()
         query_b_join_field_name = "join_{}".format(join_id)
-        target_fields = [{
-            "name": query_b_join_field_name,
-            "data_type": "object",
-            "is_list": False,
-            "is_transient": True,
-            "expression": """
-                dataset_query(
-                    "{}",
-                    fields={},
-                    filters={},
-                    entities={},
-                    limit=1
-                ) if get(record, "{}") else {{}}
-            """.format(query_b._dataset_id,
-                       [f.name for f in query_b_fields] + [key_b],
-                       filters,
-                       query_params.get('entities'),
-                       key_b)
-        }]
+        target_fields = [
+            {
+                "name": query_b_join_field_name,
+                "data_type": "object",
+                "is_list": True,
+                "is_transient": True,
+                "expression": """
+                    dataset_query(
+                        "{}",
+                        fields={},
+                        filters={},
+                        entities={},
+                        limit={}
+                    ) if get(record, "{}") else {{}}
+                """.format(query_b._dataset_id,
+                           [f.name for f in query_b_fields] + [key_b],
+                           filters,
+                           query_params.get('entities'),
+                           query_b._limit,
```

Comment:
I don't know if this is really correct to use the query b limit for this purpose. We can set this to some large limit, like 100000.

Reply:
This is necessary, let's assume that dataset B has 1m rows and query B is created against dataset B but only taking into account the first 1k rows. If we have query A with only one row consisting of the key which is in all 100 rows of query B, the result of the `join` (left outer join) method is 1 x 1k rows. However, without setting a limit to `query_b._limit` the result will be 1 x 100k rows, as `dataset_query` expression function returns up to 100k records, otherwise the exception will be raised as it is initially limited to that number.

So I have changed it a bit. If query B limit has been set up, that is used for limiting `dataset_query` otherwise limit is 100000.

Reply:
But the dataset_query expression will only return 1k records in that case (since dataset B only has 1k), regardless of the limit. I don't see how it would return 100k.